spiffe-helper.go

```diff
@@ -2,74 +2,114 @@ package main
 
 import (
 	"context"
+	"crypto/x509"
 	"encoding/pem"
-	"errors"
 	"fmt"
+	"github.com/apex/log"
+	"github.com/spiffe/spire/api/workload"
+	proto "github.com/spiffe/spire/proto/api/workload"
 	"io/ioutil"
+	"net"
 	"os"
 	"os/exec"
 	"os/signal"
 	"path"
 	"strings"
+	"sync"
 	"syscall"
 	"time"
-
-	"github.com/spiffe/spire/proto/api/workload"
 )
 
-// Sidecar is the component that consumes Workload API and renews certs
-type Sidecar struct {
-	config                *SidecarConfig
-	workloadClient        workload.WorkloadClient
-	workloadClientContext context.Context
-	processRunning        bool
-	process               *os.Process
+// sidecar is the component that consumes the Workload API and renews certs
+// implements the interface Sidecar
+type sidecar struct {
+	config            *SidecarConfig
+	processRunning    bool
+	process           *os.Process
+	workloadAPIClient workload.X509Client
+	mux               sync.Mutex
 }
 
+const (
+	// default timeout Duration for the workloadAPI client when the defaultTimeout
+	// is not configured in the .conf file
+	defaultTimeout = time.Duration(5 * time.Second)
+
+	certsFileMode = os.FileMode(0644)
+)
+
 // NewSidecar creates a new sidecar
-func NewSidecar(workloadClientContext context.Context, config *SidecarConfig, workloadClient workload.WorkloadClient) *Sidecar {
-	return &Sidecar{
-		workloadClientContext: workloadClientContext,
-		config:                config,
-		workloadClient:        workloadClient,
+func NewSidecar(config *SidecarConfig) (*sidecar, error) {
+	timeout, err := getTimeout(config)
+	if err != nil {
+		return nil, err
 	}
+
+	return &sidecar{
+		config:            config,
+		workloadAPIClient: newWorkloadAPIClient(config.AgentAddress, timeout),
+	}, nil
 }
 
 // RunDaemon starts the main loop
-func (s *Sidecar) RunDaemon() error {
+// Starts the workload API client to listen for new SVID updates
+// When a new SVID is received on the updateChan, the SVID certificates
+// are stored in disk and a restart signal is sent to the proxy's process
+func (s *sidecar) RunDaemon(ctx context.Context) error {
 	// Create channel for interrupt signal
 	interrupt := make(chan os.Signal, 1)
+	errorChan := make(chan error)
```

Comment:
this channel needs to be buffered. otherwise if Start() happens to return an error after the for-select loop has exited (but before s.workloadAPIClient.Stop() has been called) the goroutine will block.